lib/vrl/parser/src/template_string.rs

```diff
@@ -0,0 +1,76 @@
+use diagnostic::Span;
+use std::fmt;
+
+use crate::ast::{Expr, Ident, Literal::RawString, Node, Op, Opcode};
+
+#[derive(Clone, PartialEq, Eq, Ord, PartialOrd, Debug, Hash)]
+pub struct TemplateString(pub Vec<StringSegment>);
+
+impl TemplateString {
+    /// Rewrites the ast for the template string to be a series of string concatenations
+    pub fn rewrite(&self) -> Expr {
+        self.0
+            .iter()
+            .map(|node| -> Expr {
+                match node {
+                    StringSegment::Literal(s) => {
+                        Expr::Literal(Node::new(diagnostic::Span::default(), RawString(s.clone())))
+                    }
+                    StringSegment::Template(s, span) => {
+                        Expr::Variable(Node::new(*span, Ident::new(s)))
+                    }
+                }
+            })
+            .reduce(|accum, item| {
+                Expr::Op(Node::new(
+                    diagnostic::Span::default(),
+                    Op(
+                        Box::new(Node::new(diagnostic::Span::default(), accum)),
+                        Node::new(diagnostic::Span::default(), Opcode::Add),
+                        Box::new(Node::new(diagnostic::Span::default(), item)),
+                    ),
+                ))
+            })
+            .unwrap_or_else(|| {
+                Expr::Literal(Node::new(
+                    diagnostic::Span::default(),
+                    RawString("".to_string()),
+                ))
+            })
+    }
+
+    /// If the template string is just a single literal string return that string
+    /// as we can just represent it in the ast as a single literal, otherwise return
+    /// None as we will need to rewrite it into an expression.
+    pub fn literal_string(&self) -> Option<String> {
+        match self.0.as_slice() {
+            [StringSegment::Literal(s)] => Some(s.clone()),
+            _ => None,
+        }
+    }
+}
```

Comment:
A few notes w.r.t. conventions:

- I would suggest leaving the `clone` out of here, and leaving that as a decision of the caller.
- When doing so, renaming this to `as_literal_string` fits better, as it's a pattern promoted by Rust itself, and one we also heavily use in the `Value` type.

At the call-site, it'd become `template.as_literal_string().cloned()`.